src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -277,10 +277,12 @@ private string GenerateForTypeWithUnknownConverter(TypeGenerationSpec typeMetada
                 string typeCompilableName = typeMetadata.TypeRef;
                 string typeFriendlyName = typeMetadata.TypeInfoPropertyName;
 
-                StringBuilder sb = new();
+                string metadataInitSource;
 
                 // TODO (https://github.com/dotnet/runtime/issues/52218): consider moving this verification source to common helper.
-                string metadataInitSource = $@"{JsonConverterTypeRef} converter = {typeMetadata.ConverterInstantiationLogic};
+                if (typeMetadata.IsValueType)
+                {
+                    metadataInitSource = $@"{JsonConverterTypeRef} converter = {typeMetadata.ConverterInstantiationLogic};
                     {TypeTypeRef} typeToConvert = typeof({typeCompilableName});
                     if (!converter.CanConvert(typeToConvert))
                     {{
```

Comment:
Given that there is a some amount of duplication between the ref type and value type generated code, is there any way we could splice in the nullable check, i.e some thing like
```csharp
{!typeof(T).IsValueType ? "" : EmitNullableTypeHandling()}
```
?

Reply:
That is a fair suggestion, and I'll leave it up to @layomia to change it if desired. However, in this case I think it is more readable\understandable to have separate `if`s since the amount of emit duplication is only 5 LOC or so.